include/triton/backend/device_memory_tracker.h

```diff
@@ -0,0 +1,287 @@
+      return *this;
+    }
+
+    TRITONSERVER_Error* SerializeToBufferAttributes(
+      TRITONSERVER_BufferAttributes** usage, int32_t* usage_size)
+    {
+      int32_t usage_idx = 0;
+
+      // Define lambda to convert an vector of memory usage of the same type of
+      // device into buffer attributes and set in 'usage'
+      auto set_attributes_for_device_fn = [&](const std::vector<int64_t>& devices, const TRITONSERVER_MemoryType mem_type) -> TRITONSERVER_Error* {
+        for (size_t idx=0; idx < devices.size(); ++idx) {
+          // skip if no allocation
+          if (devices[idx] == 0) {
+            continue;
+          }
+          // there is space in usage array 
+          if (usage_idx < *usage_size) {
+            auto entry = usage[usage_idx];
+
+            RETURN_IF_ERROR(TRITONSERVER_BufferAttributesSetMemoryType(entry, mem_type));
+            RETURN_IF_ERROR(TRITONSERVER_BufferAttributesSetMemoryTypeId(entry, idx));
+            RETURN_IF_ERROR(TRITONSERVER_BufferAttributesSetByteSize(entry, devices[idx]));
+
+            ++usage_idx;
+          } else {
+            *usage_size = -1;
+            return TRITONSERVER_ErrorNew(TRITONSERVER_ERROR_INVALID_ARG, "More entries needed to store memory usage");
+          }
+        }
+        return nullptr;  // success
+      };
+
+      RETURN_IF_ERROR(set_attributes_for_device_fn(system_byte_size_, TRITONSERVER_MEMORY_CPU));
+      RETURN_IF_ERROR(set_attributes_for_device_fn(pinned_byte_size_, TRITONSERVER_MEMORY_CPU_PINNED));
+      RETURN_IF_ERROR(set_attributes_for_device_fn(cuda_byte_size_, TRITONSERVER_MEMORY_GPU));
+
+      *usage_size = usage_idx;
+      return nullptr;
+    }
+
+    // Byte size of allocated memory tracked,
+    // 'system_byte_size_' is likely to be empty as system memory allocation
+    // is not controlled by CUDA driver. But keeping it for completeness.
+    std::vector<int64_t> system_byte_size_{0};
+    std::vector<int64_t> pinned_byte_size_{0};
+    std::vector<int64_t> cuda_byte_size_{0};
+    bool tracked_{false};
+
+    TRITONBACKEND_CuptiTracker cupti_tracker_;
+  };
+
+  // Simple scope guard to make sure memory usage is untracked without coupling
+  // with MemoryUsage lifecycle
+  struct ScopeGuard {
+    ScopeGuard(MemoryUsage* usage) : usage_(usage) {}
+    ~ScopeGuard() {
+      if (usage_ && usage_->tracked_) {
+        UntrackThreadMemoryUsage(usage_);
+      }
+    }
+    MemoryUsage* usage_{nullptr};
+  };
+  
+
+#ifdef TRITON_ENABLE_GPU
+  static bool Init();
+
+  static int CudaDeviceCount();
+
+  // The memory usage will be tracked and modified until it's untracked, 'usage'
+  // must be valid and not to be modified externally until untrack is called.
+  // Currently can distinguish activity by correlation id which is thread
+  // specific, which implies that there will be mssing records if tracking
+  // region switching threads to handle other activities.
+  static void TrackThreadMemoryUsage(MemoryUsage* usage);
```

Comment:
Should this be the C struct? and not the C++ ? TRITONBACKEND_CuptiTracker_t

Also maybe we could rename to TRITONBACKEND_MemoryUsageTracker_t - to avoid explicitly referencing Cupti

Reply:
The C struct is what actually being passed to the CUPTI callback (user really shouldn't interact directly), `MemoryUsage` and `DeviceMemoryTracker` is the C++ class / struct that the user should interacts with.

I could be wrong as I am not an expert on the subject, but I don't think we can get rid of the C struct for ABI compatibility reason, the actual CUPTI callback and the structure used in the callback is C to ensure the (de facto) ABI is the same. Note that the ABI concern comes into the picture as CUPTI may have only one callback register (i.e. from backend A) and used by all backends which can be compiled with different C++ compiler.

Reply:
> The C struct is what actually being passed to the CUPTI callback (user really shouldn't interact directly), `MemoryUsage` and `DeviceMemoryTracker` is the C++ class / struct that the user should interacts with.
> 
> I could be wrong as I am not an expert on the subject, but I don't think we can get rid of the C struct for ABI compatibility reason, the actual CUPTI callback and the structure used in the callback is C to ensure the (de facto) ABI is the same. Note that the ABI concern comes into the picture as CUPTI may have only one callback register (i.e. from backend A) and used by all backends which can be compiled with different C++ compiler.

The MemoryUsage class though is transparent to CUPTI (i.e. it is just the external correlation ID). It doesn't dereference it. As the callback is in our C++ code - then I think we would be safe to get rid of it. However, I thought the backend API is C - so thought the the backend utilities would be C as well. -

Reply:
I see that that each backend statically links to the backend utils library - but this is the first case where we have an extern global - I'm not sure in this case exactly how things will resolve. I don't believe however that the C struct itself will handle ABI between backends. Will discuss offline.

Reply:
resolved in offline discussion - C struct used here to ensure any memory tracker singleton can be used independent of compiler